### PR TITLE
Feeler connection

### PR DESCRIPTION
**Pull request by jjyr**
Use a temporary workaround to registry feeler connection protocol, should refactor after removing CKBProtocolHandler.

### CODE REVIEW

**Comment by jjyr** [comment]
bors try

File: network/src/network.rs [hunk 1]
```diff
@@ -315,39 +297,59 @@ impl Network {
...
+            protocols.extend(
+                // TODO remove the filter after refactor
+                self.ckb_protocols
+                    .iter()
+                    .map(|p| p.0.id())
+                    .filter(|id| *id != FEELER_PROTOCOL_ID)
+                    .collect::<Vec<_>>(),
+            );
+            DialProtocol::Multi(protocols)
+        })
```
Comment (TheWaWaR) [inline network/src/network.rs]: Have you tested in your machine? This dial will not work out. We have more protocols than those 4.

Comment (jjyr) [reply network/src/network.rs]: This code dial more than 4 protocols on my machine.

**Comment by jjyr** [comment]
bors try-
bors try

**Comment by doitian** [comment]
It seems the node stuck in the integration test.

File: network/src/network.rs [hunk 2]
```diff
@@ -315,39 +297,59 @@ impl Network {
...
+        let p2p_phase =
+            Multihash::from_bytes(expected_peer_id.as_bytes().to_vec()).map_err(|_| {
+                Error::P2P(format!("peer_id is invalid {:?}", expected_peer_id).to_string())
+            })?;
+        addr.append(multiaddr::Protocol::P2p(p2p_phase));
+        self.p2p_control
+            .write()
+            .dial(addr, dial_protocol)
+            .map_err(|_| Error::P2P("dial error".to_string()))
+    }
```
Comment (TheWaWaR) [inline network/src/network.rs]: Assume I got a peer address, and it's not include a `/p2p` section, than I can't dial it?

File: network/src/network.rs [hunk 3]
```diff
@@ -315,39 +297,59 @@ impl Network {
...
+            DialProtocol::Multi(protocols)
+        })
+    }
+
+    pub fn dial_feeler(&self, expected_peer_id: &PeerId, addr: Multiaddr) -> Result<(), Error> {
+        self.inner_dial(
+            expected_peer_id,
+            addr,
+            DialProtocol::Single(FEELER_PROTOCOL_ID),
+        )
```
Comment (TheWaWaR) [inline network/src/network.rs]: Where is the code use this function?

File: network/src/network.rs (hunk 2 above)
Comment (TheWaWaR) [reply network/src/network.rs]: And very likely one will update his private key, than `/p2p` section is invalid, how to deal with it?

Comment (jjyr) [reply network/src/network.rs]: CKB network security and scoring base on `peer_id` and `address`(RFC0007), we must enforce connection encrypt.

Comment (jjyr) [reply network/src/network.rs]: If a peer updated with a new key, In the current implementation, the address(without p2p phase) will consider as invalid, until node rediscovery the new address(with new p2p phase).

File: network/src/network.rs (hunk 3 above)
Comment (jjyr) [reply network/src/network.rs]: For feeler connection, see `protocol/feeler`

File: network/src/network.rs (hunk 2 above)
Comment (TheWaWaR) [reply network/src/network.rs]: `connection encrypt` is not require peer_id, you just dial, `secio` will know target peer's peer_id.

Comment (TheWaWaR) [reply network/src/network.rs]: And, `/p2p` section required in `bootnodes` is not user friendly.

File: network/src/network.rs (hunk 3 above)
Comment (TheWaWaR) [reply network/src/network.rs]: You can try `Ctrl + F` on this page, than type: `dial_feeler`

File: network/src/network.rs (hunk 2 above)
Comment (jjyr) [reply network/src/network.rs]: Of cause we need `/p2p/` phase to make sure bootnodes' ips not controlled be bad crackers. Pls see RFC0007 first.

File: network/src/network.rs (hunk 3 above)
Comment (jjyr) [reply network/src/network.rs]: My mistake, fixed.

File: network/src/network.rs (hunk 2 above)
Comment (TheWaWaR) [reply network/src/network.rs]: Yes, `bootnode` must require peer_id, but RFC0007 is not show how peer_id is contribute to the security.

As a normal user, he may change his private key from time to time, but he always use the same address. I think it's a OK behavior, but will he be banned?

Comment (jjyr) [reply network/src/network.rs]: He **will not be banned**.

The reason we need to deny the old address is as same as the `bootnode` security, a user with good behaviors may earn more scores in his peers' side, once the private key changed, we can't prove the peer's IP is controlled by the same good people or a bad cracker, we must reset the trust(score) to this address.

A normal user may change his key once or twice, but not every time he start the node.

By the way, even the user do it, node outbound connections work as well; only inbound peers will not trust the old score, these peers consider the node as a new node.

There's nothing about * user-friendly*; there's a high probability the user can't feel different after reset key.

**Comment by jjyr** [comment]
bors try

**Pull request merged**